kuka_kss_hw_interface/krl/KSS_85/Program/ROS/ros_krlptp.src

```diff
@@ -0,0 +1,74 @@
+&ACCESS RVP
+&REL 74
+&PARAM DISKPATH = KRC:\R1\Program\ROS
+DEF ros_krlptp ( )
+   ; Software License Agreement (BSD License)
+   ;
+   ; Copyright (c) 2022, Kuka Robotics Corp
+   ; All rights reserved.
+   ;
+   ; Redistribution and use in source and binary forms, with or without
+   ; modification, are permitted provided that the following conditions are met:
+   ;
+   ;      * Redistributions of source code must retain the above copyright
+   ;        notice, this list of conditions and the following disclaimer.
+   ;      * Redistributions in binary form must reproduce the above copyright
+   ;        notice, this list of conditions and the following disclaimer in the
+   ;        documentation and/or other materials provided with the distribution.
+   ;      * Neither the name of the copyright holder, nor the names of its
+   ;        contributors may be used to endorse or promote products derived
+   ;        from this software without specific prior written permission.
+   ;
+   ; THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+   ; AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+   ; IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+   ; ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+   ; LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+   ; CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+   ; SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+   ; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+   ; CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+   ; ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+   ; POSSIBILITY OF SUCH DAMAGE.
+
+   ; Author: Pat Duda <pat.duda@kuka.com>
+
+;FOLD INI
+  ;FOLD BASISTECH INI
+    GLOBAL INTERRUPT DECL 3 WHEN $STOPMESS==TRUE DO IR_STOPM ( )
+    INTERRUPT ON 3 
+    BAS (#INITMOV,0 )
+  ;ENDFOLD (BASISTECH INI)
+  ;FOLD USER INI
+    ;Make your modifications here
+
+  ;ENDFOLD (USER INI)
+  ;ENDFOLD (INI)
+
+   ROS_joint_pos_tgt = $AXIS_ACT_MEAS
+   nextMove = $AXIS_ACT_MEAS
+   moveTime = 0.200
+   ptp ROS_joint_pos_tgt
+   ROS_Runtime_Init()
+   ; Setup default motion params
+   ; VEL & ACC may be reduced to give smooth motion at the default data rate
+   ; For faster performance, these may also be increased
+   ; Small/fast robots may need lower values
+   ROS_Pdat={VEL 50.000,ACC 50.000,APO_DIST 20.000,GEAR_JERK 100.000,EXAX_IGN 0}
+   PDAT_ACT=ROS_Pdat
+   BAS(#PTP_DAT)
+   $ADVANCE = 2
+   ; Loop forever
+   LOOP
+      movesQueued = ROS_GetNextMoveFIFO(nextMove, moveTime)
+      CONTINUE
+      IF (movesQueued > 0) THEN
+         PTP nextMove C_PTP
```

Comment:
Does that imply that for dense trajectories the maximum achievable speed will be limited by the distance between points? Ie so that a full stop would be possible at any time if no further points where to follow

We use splines here for our internal driver and love this feature of Kuka

Reply:
Yes.  The legacy motion planner will limit velocity if point spacing is very close.
I also found that the EKI update rate of 20Hz is a bit slow for smaller robots.  The robot executes the planned moves faster than the update rate and is starved of data so it plans a stop.  This results in jerky motion.
For your spline motions, to use the single line SPTP or do you download a trajectory as a spline block?  Do you suggest we change the PTP to SPTP in ros_krlptp.src?
I experimented with a spline TIME_BLOCK using the moveTime parameter.  I couldn't get this to provide continuous motion.
The architecture of this new driver enables different operating mode programs so we could assign different programs to each use a different motion strategy.  This might be helpful if there is not a "best" solution to fit all cases.

Reply:
We use several spline_ptp blocks of varying point size/lenghts that we concatenate. Unfortunately we found no way to use them with a variable length of points directly. You can visually tell where two blocks are connected. We try to guess a velocity override based on the ratio of velocity to max velocity, this is not accurate at all but sufficient for bin picking.

I don't know if switching this line to SPTP is advisable or not, you got the @ kuka.com address :wink: